test/compressionworker.ts

```diff
@@ -0,0 +1,222 @@
+
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    addEventMessage.event = secondMockEvent;
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(forceUploadMessage));
+    scheduleTestFailureTimeout();
+
+    let handlerInvocationCount = 0;
+    let payloads: any[] = [];
+    function messageHandler(message: IWorkerMessage) {
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      payloads.push(payload);
+      handlerInvocationCount++;
+      if (handlerInvocationCount > 1) {
+        performAssertions();
+      }
+    }
+
+    function performAssertions() {
+      assert.equal(payloads.length, 2);
+      assert.equal(payloads[0].events.length, 1);
+      assert.equal(payloads[0].events[0].type, firstMockEventName);
+      assert.equal(payloads[1].events.length, 1);
+      assert.equal(payloads[1].events[0].type, secondMockEventName);
+      done();
+    }
+  });
+
+  it("validates that single event is sent when its own length is above the limit", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let mockEvent = getMockEvent();
+    let mockEventData = Array(Math.round(config.batchLimit + 1)).join("1");
+    mockEvent.state = { data: mockEventData };
+    let addEventMessage = createAddEventMessage(mockEvent);
+
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    scheduleTestFailureTimeout();
+
+    function messageHandler(message: IWorkerMessage) {
+      assert.equal(message.type, WorkerMessageType.Upload);
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      assert.equal(payload.events.length, 1);
+      assert.equal(payload.events[0].type, MockEventName);
+      done();
+    }
+  });
+
+  it("validates that queued payload is sent on terminate", (done: DoneFn) => {
+    let worker = createTestWorker();
+    let mockEvent = getMockEvent();
+    let addEventMessage = createAddEventMessage(mockEvent);
+    let terminateMsg: ITimestampedWorkerMessage = {
+      type: WorkerMessageType.Terminate,
+      time: new Date().getTime()
+    };
+
+    assert.equal(workerMessages.length, 0);
+    messageHandlers.push(messageHandler);
+    worker.postMessage(JSON.stringify(addEventMessage));
+    worker.postMessage(JSON.stringify(terminateMsg));
+    scheduleTestFailureTimeout();
+
+    function messageHandler(message: IWorkerMessage) {
+      assert.equal(message.type, WorkerMessageType.Upload);
+      let uploadMessage = message as IUploadMessage;
+      let payload = JSON.parse(uploadMessage.rawData);
+      assert.equal(payload.events.length, 1);
+      assert.equal(payload.events[0].type, MockEventName);
+      done();
+    }
+  });
+
+  it("validates that payloads consisting of a single XHR error event are not uploaded", (done: DoneFn) => {
```

Comment:
Should these never be uploaded? Doesn't it depend on state?

Reply:
This is an edge case. During normal Clarity runtime we don't sent a single XHR event because there is an infinite loop possibility:

SendEvent -> FailToDeliver -> Generate XHR error event -> Send XHR error event -> FailToDeliver -> Generate XHR error event -> Send XHR error event...etc.

On teardown, however, single XHR error event will be uploaded synchronously from core. I can add an extra test for that as well.

---
In reply to: [136876622](https://github.com/Microsoft/clarity-js/pull/28#discussion_r136876622) [](ancestors = 136876622)